### DISCUSSION TITLE
Keeping underlying Fortran-contiguous ndarray buffer valid on return from C++ code

### DISCUSSION THREAD

**Discussion by Chronum94** [wjakob/nanobind#313] [Q&A] [answered]
Hello,

Great library, pleasure to use it. 

I'm currently creating a Python API to a C++ library which has a pattern that like so:

```cpp
.def_static(
          "from_ndarray",
          [](nb::ndarray<double, nb::device::cpu> &a)
          {
            int local_m = a.shape(0);
            int local_n = a.shape(1);

            double *local_tmp = new double[local_m * local_n];
            std::copy(a.data(), a.data() + local_m * local_n, local_tmp);

            return MatrixObject::from_ndarray(
                local_tmp);
          })
```

I am currently copying the ndarray's contents into a raw pointer but if possible, I'd like to avoid that copy operation (it's not too bad, but I'd like to avoid it).

The Numpy arrays that are fed into this C++ code will definitely remain valid (non-GC'd) for the lifetime of the returned MatrixObject. The MatrixObject requires that the memory at a.data() remains valid for the lifetime of the MatrixObject. Is there a way to do this without the copy op?

I don't know if there's any more information I can provide but I'd be more than happy to provide as much detail as I can.

**Comment by wjakob** [comment, marked as answer]
Two ways to do this: modify your `MatrixObject` to stash the `nb::ndarray` somewhere. Or use a `keep_alive` call annotation to ensure that the argument (`a`) is kept alive at least as long as the return value of the function.

**Reply by Chronum94** [reply]
I'm accepting this as the answer since a `nb::keep_alive<0, 1>()` did do the needful. The `np.asfortranarray` situation later on in the thread is mostly a different issue.

**Comment by Chronum94** [comment]
The above example is with a `.def_static` but I've been trying the `keep_alive` annotation by making a standalone function. It looks like (slightly simplified but the core logic is here) so:

```cpp
m.def(
    "matrix_from_ndarray",
    [](nb::ndarray<double, nb::device::cpu> a)
    {
        return MatrixObject::from_ndarray((double *)a.data());
    }, nb::keep_alive<0, 1>());
```

My understanding from the `keep_alive` documentation is that this means that as long as the returned `MatrixObject` (the 0) is alive, `a` (the 1) is kept alive. But this isn't working as far as I can tell. I still get segfaults, nondeterministic corruption of the numerical data.

I think this is from the return value policy (https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4N8nanobind9rv_policy9automaticE) being automatically set to `move` since I am returning an rvalue, although I could be wrong. But if the rv policy is `move`, then I don't know what the behaviour of `keep_alive` would be. I've tried changing the `rv_policy` to `reference_internal` but that also gives me a segfault.

Still very new to the library and a bit rusty with C++ memory management so I really appreciate the help through this.

**Reply by wjakob** [reply]
`reference_internal` is for a different use case: to keep a field returned by a method alive, but you are looking at static functions. The `nb::keep_alive<0, 1>` annotation looks good, I am not sure why it is not working. Can you make a self-contained reproducer so that I can investigate?

**Reply by Chronum94** [reply]
Hi,

I tried making a minimum working example, and it worked exactly as you said it would. Which made me look into this a bit more:

In my example code so far, I have been simplifying a few things. There is an `f_contig` flag that I didn't mention because I did not think it was relevant. But the actual (static) method looks like so:

```cpp
.def_static(
  "from_ndarray",
  [](nb::ndarray<double, nb::device::cpu, nb::f_contig> a)
  {
    return MatrixObject::from_ndarray((double *)a.data());
  })
```

On the python side, I try the following calls (`a` is a numpy array):

1. `MatrixObject.from_ndarray(a)`: Invalid.
2. `MatrixObject.from_ndarray(a.copy())`: Invalid.
3. `MatrixObject.from_ndarray(np.asfortranarray(a))`: **Valid.**
4. `MatrixObject.from_ndarray(np.asfortranarray(a.copy()))`: **Valid.**
5. `MatrixObject.from_ndarray(np.asfortranarray(a).copy())`: Invalid.

So on the one hand, I'm quite happy that I now don't have to do the copy. On the other hand, I don't understand why `nb:f_contig` gives the correct 1D array element ordering in both cases (whether I use `np.asfortranarray` or not), but in one case requires a copy of the data.

**Reply by wjakob** [reply]
I know what's happening here. nanobind is making an implicit conversion to turn a standard NumPy array (with C-style ordering) into one that conforms to your `ndarray` specification (with F-style ordering). The `keep_alive` annotation keeps the original array function input alive, but not the implicitly converted one. That's a limitation of the current nanobind design, but a quite confusing one I must say.

**Reply by Chronum94** [reply]
Implicit conversion strikes again. Nevertheless, thanks so much for the help with all of this. Both the `keep_alive` and the subtle F-contig issue. This has likely saved me weeks.

**Reply by wjakob** [reply]
This is still very bad. I want to fix it, stay tuned..

**Reply by wjakob** [reply]
@Chronum94 could I ask you for a favor?

Could you test your codebase (with the crashing/undefined behavior/inconsistencies) with the nanobind `master` branch, specifically commits `f0a9ebd9cd384ac554312247526b120102563e53` and `0a1ccfebef3ba107325ebad2a168c85a5283d837`? I made some changes just now that perhaps address the issue.

If my changes work, then I would expect the first commit to manifest the problems you reported, while the second one fixes them. (That assumes that the undefined behavior is consistent enough to easily detect).

**Reply by wjakob** [reply]
Hm never mind, I am not happy with those changes yet and will revert them.

**Reply by Chronum94** [reply]
Just got around to running a preliminary test of this. The solution indeed works as intended. I now no longer have to supply an `asfortranarray` on the Numpy side of things, and the `keep_alive` works as intended!

I'll run some more tests either later this evening or this upcoming week (long weekend in the USA so likely on Tuesday). But for now the changes appear to be working as intended.

**Reply by Chronum94** [reply]
I'm more than happy to test things out as you get them. The changes in the second commit above did work exactly as intended from the tests I was running earlier. I can test again whenever you have changes you're happy with.

**Reply by wjakob** [reply]
I did some more work on this -- could you try again with `master`?

**Reply by Chronum94** [reply]
Ran some quick tests, seems to be working as intended on `master` (no `np.asfortranarray` on the Python side, visibly correct behaviour on the C++ side). Should be able to check more comprehensively on Monday.

**Reply by wjakob** [reply]
Great, thank you -- I'll wait to for your thumbs up and then plan to ship this in a new release soon,

**Reply by Chronum94** [reply]
```
In file included from /miniconda3/envs/devenv/lib/python3.10/site-packages/nanobind/include/nanobind/nanobind.h:43,
                 from /miniconda3/envs/devenv/lib/python3.10/site-packages/nanobind/src/nb_internals.cpp:10:
/miniconda3/envs/devenv/lib/python3.10/site-packages/nanobind/include/nanobind/nb_lib.h:61:10: error: 'bool nanobind::detail::cleanup_list::used()' cannot be overloaded with 'bool nanobind::detail::cleanup_list::used()'
   61 |     bool used() { return m_size != 1; }
      |          ^~~~
/miniconda3/envs/devenv/lib/python3.10/site-packages/nanobind/include/nanobind/nb_lib.h:53:20: note: previous declaration 'bool nanobind::detail::cleanup_list::used()'
   53 |     NB_INLINE bool used() { return m_size != 0; }
      |                    ^~~~
/miniconda3/envs/devenv/lib/python3.10/site-packages/nanobind/include/nanobind/nb_lib.h: In member function 'PyObject* nanobind::detail::cleanup_list::operator[](size_t) const':
/miniconda3/envs/devenv/lib/python3.10/site-packages/nanobind/include/nanobind/nb_lib.h:67:67: error: cannot convert 'nanobind::detail::cleanup_list::entry' to 'PyObject*' {aka '_object*'} in return
   67 |     PyObject *operator[](size_t index) const { return m_data[index]; }
      |                                                       ~~~~~~~~~~~~^
      |                                                                   |
      |                                                                   nanobind::detail::cleanup_list::entry
make[2]: *** [CMakeFiles/nanobind-static.dir/build.make:76: CMakeFiles/nanobind-static.dir/miniconda3/envs/devenv/lib/python3.10/site-packages/nanobind/src/nb_internals.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:110: CMakeFiles/nanobind-static.dir/all] Error 2
```

Reran some tests and I get this little issue here with master. I've redacted parts of the paths, but that's just the home directory part of the full path.

**Reply by wjakob** [reply]
Is it possible that you have conflicting versions of nanobind installed? For example, a system/Pip version and a local one? The line numbers you posted don't match up with the current version on `master`.

**Reply by wjakob** [reply]
Maybe there are some leftover changes on your end from the previous (retracted) change I had made a while back?

**Reply by Chronum94** [reply]
Hmm it's possible I'll have a check and make sure things agree. I did manually remove a nanobind from pip before using master, and `python -m nanobind --version` does give me 1.6.2, but I'll double check and clear this up.

**Reply by Chronum94** [reply]
The problem seems to be back:

Pasting a MWE here:

`mwe.cpp`
```cpp
#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>

namespace nb = nanobind;

class A // begin declaration of the class
{
private:
    A(double *a, size_t n)
    {
        a_ = a;
        n_ = n;
    }

public:
    double *a_;
    size_t n_;
    static A from_raw(double *a, size_t n)
    {
        return A(a, n);
    }
};

NB_MODULE(mwe, m)
{
    nb::class_<A>(m, "A")
        .def_static(
            "from_ndarray",
            [](nb::ndarray<double, nb::device::cpu, nb::f_contig> a)
            {
                return A::from_raw((double *)a.data(), (size_t)a.shape(0));
            })
        .def(
            "numpy",
            [](A a)
            {
                size_t shape[2] = {a.n_, a.n_};

                return nb::ndarray<nb::numpy, double, nb::shape<nb::any>, nb::f_contig>(a.a_, 2, shape);
            },
            nb::keep_alive<0, 1>());
}
```

`test.py`
```python
from mwe import A
import numpy as np


np.set_printoptions(precision=4, linewidth=250, formatter={'float': lambda x: f"{x:>10.4f}"})

N = 1000
np.random.seed(24365)
a = np.random.rand(N, N)
b = A.from_ndarray(a)
print(a)
print(b.numpy())

print(np.allclose(a, b.numpy().T))
```
`CMakeLists.txt`
```cmake
cmake_minimum_required(VERSION 3.15..3.27)
project(my_project) # Replace 'my_project' with the name of your project
# cmake_minimum_required(VERSION 3.15...3.27)
find_package(Python 3.10.12 COMPONENTS Interpreter Development.Module REQUIRED)
if (NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
  set(CMAKE_BUILD_TYPE Release CACHE STRING "Choose the type of build." FORCE)
  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
endif()

execute_process(
 COMMAND "${Python_EXECUTABLE}" -m nanobind --cmake_dir
 OUTPUT_STRIP_TRAILING_WHITESPACE OUTPUT_VARIABLE NB_DIR)
list(APPEND CMAKE_PREFIX_PATH "${NB_DIR}")
find_package(nanobind CONFIG REQUIRED)

nanobind_add_module(mwe NOSTRIP mwe.cpp)
```

Steps to reproduce:


1. Set `N=10` in `test.py` and run, should output **exactly** this:
```
[[    0.8303     0.4806     0.2312     0.5671     0.8168     0.1924     0.9145     0.5729     0.2217     0.1311]
 [    0.9006     0.2502     0.6532     0.1115     0.7973     0.4337     0.7268     0.7043     0.9412     0.3475]
 [    0.2424     0.1325     0.0887     0.7227     0.8628     0.9500     0.0957     0.2697     0.0214     0.9608]
 [    0.8875     0.1085     0.3836     0.9166     0.1517     0.2555     0.2762     0.4151     0.9531     0.4796]
 [    0.9369     0.4754     0.0300     0.0155     0.7501     0.3923     0.6552     0.2583     0.4389     0.6955]
 [    0.8376     0.8764     0.7786     0.6817     0.4649     0.1332     0.2816     0.0540     0.7071     0.5322]
 [    0.3756     0.5338     0.8753     0.4635     0.5019     0.1981     0.6977     0.8027     0.7241     0.3669]
 [    0.2700     0.2520     0.5283     0.2112     0.6829     0.2557     0.5499     0.1072     0.7430     0.5279]
 [    0.8371     0.4252     0.6877     0.9882     0.6469     0.8304     0.8468     0.8140     0.1524     0.4194]
 [    0.7763     0.4885     0.7786     0.3065     0.9863     0.6828     0.6155     0.9877     0.0428     0.1670]]
[[    0.8303     0.9006     0.2424     0.8875     0.9369     0.8376     0.3756     0.2700     0.8371     0.7763]
 [    0.4806     0.2502     0.1325     0.1085     0.4754     0.8764     0.5338     0.2520     0.4252     0.4885]
 [    0.2312     0.6532     0.0887     0.3836     0.0300     0.7786     0.8753     0.5283     0.6877     0.7786]
 [    0.5671     0.1115     0.7227     0.9166     0.0155     0.6817     0.4635     0.2112     0.9882     0.3065]
 [    0.8168     0.7973     0.8628     0.1517     0.7501     0.4649     0.5019     0.6829     0.6469     0.9863]
 [    0.1924     0.4337     0.9500     0.2555     0.3923     0.1332     0.1981     0.2557     0.8304     0.6828]
 [    0.9145     0.7268     0.0957     0.2762     0.6552     0.2816     0.6977     0.5499     0.8468     0.6155]
 [    0.5729     0.7043     0.2697     0.4151     0.2583     0.0540     0.8027     0.1072     0.8140     0.9877]
 [    0.2217     0.9412     0.0214     0.9531     0.4389     0.7071     0.7241     0.7430     0.1524     0.0428]
 [    0.1311     0.3475     0.9608     0.4796     0.6955     0.5322     0.3669     0.5279     0.4194     0.1670]]
```
2. Set `N=1000` in `test.py` -do-:
```
[[    0.8303     0.4806     0.2312 ...     0.4643     0.5050     0.0717]
 [    0.0512     0.3386     0.4343 ...     0.0762     0.2647     0.5382]
 [    0.6355     0.0298     0.0473 ...     0.6699     0.3410     0.7473]
 ...
 [    0.4101     0.3293     0.2514 ...     0.5839     0.6020     0.3755]
 [    0.1256     0.2223     0.4554 ...     0.6677     0.5997     0.3520]
 [    0.5353     0.8823     0.3914 ...     0.6054     0.0165     0.0949]]
Segmentation fault (core dumped)
```

The problems:

1. The segfault at N=1000. This is the problem we are dealing with currently.
2. The final array that I get back is transposed compared to what went in. This is less important since it might just be me not knowing enough about the ndarray annotations to return a properly strided/transposed array.

Please let me know if there is anything more I can provide. Hopefully this helps in reproducing the issue.

**Reply by wjakob** [reply]
Here is some advice to change your code:

```cpp
    nb::class_<A>(m, "A")
        .def_static(
            "from_ndarray",
            [](nb::ndarray<double, nb::device::cpu, nb::f_contig> a)
            {
                return A::from_raw((double *)a.data(), (size_t)a.shape(0));
            },
            // Part 1: the created `A` instance should keep the input `a` alive.
            nb::keep_alive</* Nurse = */ 0, /* Patient */ 1>())

        .def(
            "numpy",
            [](A *a)
            {
                size_t shape[2] = {a->n_, a->n_};
                int64_t strides[2] = {1, (Py_ssize_t) a->n_};

                // Part 2: when returning a nd-array, you must specify which Python
                // object owns the underlying storage. It's better to handle
                // this using the specialized mechanism ('owner' field of the
                // ndarray constructor) as opposed to the generic keep_alive
                // annotation.

                // I changed the function argument to a pointer so that we can
                // find the associated Python object associated with the C++ memory
                // address.
                nb::object owner = nb::find(a);

                // When returning an nd-array, all of the annotations below
                // (except *nb::numpy*) are just decorations for the
                // docstring. In contrast to accepting an nd-array as input,
                // no conversions are performed. In other words, it's your job
                // to provide a numpy array of the promised properties. You
                // must also specify a strides vector if you use non-C strides.
                return nb::ndarray<nb::numpy, double, nb::shape<nb::any>,
                                   nb::f_contig>(a->a_, 2, shape, owner, strides);
            });
```

With these changes, I don't have the issues you mentioned with undefined behavior. Can you confirm?

**Reply by wjakob** [reply]
And here is an even better version using `nb::pointer_and_handle` that simultaneously accepts the C++ and Python version of an object to avoid the need for a subsequent ``nb::find`` call.

```cpp
        .def(
            "numpy",
            [](nb::pointer_and_handle<A> pah)
            {
                A *a = pah.p;
                nb::handle h = pah.h;

                size_t shape[2] = {a->n_, a->n_};
                int64_t strides[2] = {1, (Py_ssize_t) a->n_};

                // When returning an nd-array, all of the annotations below
                // (except *nb::numpy*) are just decorations for the
                // docstring. In contrast to accepting an nd-array as input,
                // no conversions are performed. In other words, it's your job
                // to provide a numpy array of the promised properties. You
                // must also specify a strides vector if you use non-C strides.
                return nb::ndarray<nb::numpy, double, nb::shape<nb::any>,
                                   nb::f_contig>(a->a_, 2, shape, h, strides);
            });
```

**Reply by Chronum94** [reply]
After fairly aggressive testing of the exact code I have in production (and fixing some import errors I'm not proud of), I can confirm that `master` fixes the problem.

Also really appreciate the code review of the mwe. Learning good practices/new things as I go.

I deeply appreciate all of the time you've spent on this, and I think `master` fixes the subtlety with `keep_alive`.

**Reply by wjakob** [reply]
Great, thanks for the kind words and for double-checking.

**Reply by Chronum94** [reply]
I've changed the title slightly to make the discussion somewhat easier to find in the future should anyone need it. Unless you think otherwise, we can close this discussion since all issues have been addressed.

**Reply by wjakob** [reply]
Sounds good!
